### ISSUE TITLE
Project Goal

### ISSUE THREAD

**Issue by xlu-agentic-systems** [xlu-agentic-systems/swe-benchmark#66]
# Project: Autonomy-Style Perception Evaluation System

## Goal

Build a small but realistic ML infra project that simulates part of an autonomous driving perception workflow:

```text
Driving images / frames
        ↓
PyTorch object detection model
        ↓
ONNX export
        ↓
C++ inference runner
        ↓
Evaluation metrics
        ↓
Python report/dashboard
```

The project should show:

```text
C++ systems skill
PyTorch understanding
model export/inference knowledge
benchmarking/profiling
autonomy-domain relevance
clean engineering
```

---

# Phase 0: Define the MVP

Do **not** start too big. The MVP should answer:

> Given a folder of driving images, can my system detect vehicles/pedestrians, run inference in C++, measure latency, and generate an evaluation report?

## MVP scope

Input:

```text
images/
  frame_0001.jpg
  frame_0002.jpg
  ...
```

Model output:

```json
{
  "frame_id": "frame_0001.jpg",
  "detections": [
    {
      "class": "car",
      "confidence": 0.91,
      "bbox": [x1, y1, x2, y2]
    }
  ],
  "latency_ms": 18.4
}
```

Metrics:

```text
p50 latency
p95 latency
throughput FPS
number of detections per frame
false positives per frame
precision / recall if labels are available
memory usage
```

---

# Phase 1: Set up repo structure

Use a clean structure from day one.

```text
autonomy-perception-eval/
  README.md
  CMakeLists.txt

  data/
    raw/
    processed/
    samples/

  models/
    pytorch/
    onnx/

  python/
    prepare_dataset.py
    train_or_load_model.py
    export_onnx.py
    evaluate_outputs.py
    generate_report.py
    dashboard.py

  cpp/
    inference/
      main.cpp
      OnnxDetector.h
      OnnxDetector.cpp
      ImagePreprocessor.h
      ImagePreprocessor.cpp
      Benchmark.h
      Benchmark.cpp

  configs/
    model.yaml
    eval.yaml

  outputs/
    detections/
    metrics/
    reports/

  tests/
    test_preprocessing.py
    test_metrics.py
    cpp_tests/
```

This already looks like a serious infra-style project.

---

# Phase 2: Pick dataset

Start with something small.

Recommended order:

```text
1. Use 100–500 sample driving images first
2. Add labels later
3. Then move to a real dataset like KITTI / BDD100K / nuScenes mini
```

For the first version, you can even use unlabeled images and only measure:

```text
latency
throughput
detections per frame
confidence distribution
```

Then in version 2, add labels and calculate:

```text
precision
recall
false positives
false negatives
IoU
```

## Dataset format you want internally

Normalize everything into one simple format:

```json
{
  "image_id": "frame_0001.jpg",
  "width": 1280,
  "height": 720,
  "objects": [
    {
      "class": "car",
      "bbox": [100, 120, 300, 260]
    }
  ]
}
```

Do this even if the original dataset has a different annotation format.

This shows good backend/data engineering discipline.

---

# Phase 3: PyTorch model path

You have two options.

## Option A: Fast MVP — use a pretrained model

This is better for your first version.

Use a pretrained object detection model and focus on the infra:

```text
image preprocessing
model loading
ONNX export
C++ inference
benchmarking
evaluation
reporting
```

This is closer to ML infra.

## Option B: Fine-tune later

After MVP works, fine-tune on a small driving dataset.

For example:

```text
classes:
  car
  pedestrian
  cyclist
  traffic light
  bus
  truck
```

But do not start here. Training object detectors can slow you down.

---

# Phase 4: Build PyTorch inference script first

Before C++, make Python inference work.

Create:

```text
python/train_or_load_model.py
```

Responsibilities:

```text
load model
load image
run inference
save detection output JSON
```

Example output:

```text
outputs/detections/python/frame_0001.json
outputs/detections/python/frame_0002.json
```

Success criteria:

```text
You can run inference on 100 images.
You can save detection JSON files.
You can visualize bounding boxes for a few frames.
```

---

# Phase 5: Export model to ONNX

Create:

```text
python/export_onnx.py
```

Flow:

```text
PyTorch model
        ↓
dummy input tensor
        ↓
torch.onnx.export
        ↓
models/onnx/detector.onnx
```

Your export script should print:

```text
model input shape
model output shape
ONNX file size
export success/failure
```

Success criteria:

```text
ONNX model exists.
ONNX model can run from Python using ONNX Runtime.
Python PyTorch output and ONNX output are roughly consistent.
```

This is a very important ML infra skill.

---

# Phase 6: Validate ONNX in Python

Before writing C++, validate the ONNX model using Python.

Create:

```text
python/validate_onnx.py
```

It should compare:

```text
PyTorch inference output
ONNX Runtime inference output
```

You want to check:

```text
same number of output tensors
same tensor shapes
similar confidence values
reasonable bounding boxes
```

Do not expect perfect equality because of runtime differences, but it should be close enough.

---

# Phase 7: Build C++ inference runner

This is the core portfolio value.

Create:

```text
cpp/inference/main.cpp
cpp/inference/OnnxDetector.h
cpp/inference/OnnxDetector.cpp
```

The C++ runner should do:

```text
load ONNX model
load image
resize / normalize image
run inference
postprocess boxes
write output JSON
record latency
```

Command-line interface:

```bash
./detector \
  --model models/onnx/detector.onnx \
  --input data/samples \
  --output outputs/detections/cpp \
  --batch_size 1
```

Output:

```json
{
  "frame_id": "frame_0001.jpg",
  "latency_ms": 14.8,
  "detections": [
    {
      "class_id": 2,
      "class_name": "car",
      "score": 0.93,
      "bbox": [102, 118, 296, 262]
    }
  ]
}
```

Success criteria:

```text
C++ can run the ONNX model.
C++ outputs valid JSON.
C++ latency is measured per frame.
C++ results are close to Python ONNX results.
```

---

# Phase 8: Add benchmarking

Create:

```text
cpp/inference/Benchmark.h
cpp/inference/Benchmark.cpp
```

Measure:

```text
model load time
image preprocessing time
inference time
postprocessing time
total frame time
p50 latency
p95 latency
p99 latency
throughput FPS
```

Output:

```json
{
  "num_frames": 500,
  "avg_latency_ms": 17.2,
  "p50_latency_ms": 15.8,
  "p95_latency_ms": 28.4,
  "p99_latency_ms": 35.1,
  "throughput_fps": 58.1
}
```

This is where your backend/performance background shines.

---

# Phase 9: Add evaluation metrics

Create:

```text
python/evaluate_outputs.py
```

If labels are available, compute:

```text
IoU
true positives
false positives
false negatives
precision
recall
F1
average confidence
false positives per frame
missed objects per frame
```

Basic IoU logic:

```text
predicted box vs ground-truth box
        ↓
calculate overlap
        ↓
if IoU >= 0.5 and class matches:
    true positive
else:
    false positive
```

Output:

```json
{
  "precision": 0.82,
  "recall": 0.74,
  "f1": 0.78,
  "false_positives_per_frame": 1.3,
  "missed_objects_per_frame": 0.8
}
```

Success criteria:

```text
You can compare detection output against labels.
You can identify where the model performs badly.
You can summarize accuracy and latency together.
```

---

# Phase 10: Generate report/dashboard

Create:

```text
python/generate_report.py
```

The report should include:

```text
latency summary
throughput summary
precision / recall
false positive examples
missed detection examples
confidence distribution
worst latency frames
sample visualizations
```

Output:

```text
outputs/reports/report.html
```

or

```text
outputs/reports/report.md
```

Example report sections:

```text
1. Dataset Summary
2. Model Summary
3. Inference Benchmark
4. Detection Quality
5. Failure Cases
6. Optimization Opportunities
```

This makes the project interview-friendly.

---

# Phase 11: Add visualizations

For selected frames, draw:

```text
ground-truth boxes in one style
predicted boxes in another style
confidence score
class label
```

Generate:

```text
outputs/reports/visualizations/frame_0001.jpg
```

Useful categories:

```text
best examples
worst false positives
missed pedestrians
highest latency frames
low-confidence detections
```

This makes the project much easier to explain.

---

# Phase 12: Add C++ engineering quality

This is where you make it look like a real systems project.

Add:

```text
CMake
unit tests
error handling
structured logging
config files
clean interfaces
sanitizer support
benchmark mode
```

Important C++ classes:

```cpp
class OnnxDetector {
public:
    explicit OnnxDetector(const DetectorConfig& config);
    DetectionResult run(const cv::Mat& image);
};

class ImagePreprocessor {
public:
    Tensor preprocess(const cv::Mat& image);
};

class BenchmarkRecorder {
public:
    void recordLatency(double milliseconds);
    BenchmarkSummary summarize() const;
};
```

This shows you can design maintainable C++ code, not just scripts.

---

# Phase 13: Optional advanced features

After MVP, add these one by one.

## Advanced Feature 1: Batch inference

Compare:

```text
batch size 1
batch size 4
batch size 8
```

Measure:

```text
latency per frame
throughput
memory usage
```

This teaches real inference tradeoffs.

## Advanced Feature 2: Async pipeline

Build a C++ pipeline:

```text
image loader thread
        ↓
preprocessing queue
        ↓
inference worker
        ↓
postprocessing worker
        ↓
output writer
```

This connects directly to your C++ concurrency strength.

## Advanced Feature 3: Cloud NVIDIA benchmark

Run the same ONNX model on a cloud GPU.

Compare:

```text
Mac MPS / CPU
ONNX Runtime CPU
NVIDIA CUDA
TensorRT
```

This is highly relevant to NVIDIA/Tesla-style infra roles.

## Advanced Feature 4: Scenario-level evaluation

Instead of evaluating individual frames only, evaluate sequences:

```text
Did the model consistently detect the same car across frames?
Did detection confidence drop at night?
Did pedestrian detection fail in crowded scenes?
Did latency spike on high-object-count frames?
```

This feels much more autonomy-specific.

---

# Suggested 6-week build plan

## Week 1: Dataset + Python inference

Deliverables:

```text
repo setup
sample dataset loader
pretrained PyTorch model inference
JSON output
basic visualization
```

## Week 2: ONNX export

Deliverables:

```text
ONNX export script
ONNX Runtime validation
PyTorch vs ONNX comparison
clean model config
```

## Week 3: C++ inference

Deliverables:

```text
C++ ONNX Runtime runner
OpenCV image preprocessing
JSON output
single-frame inference
```

## Week 4: Benchmarking

Deliverables:

```text
batch image inference
latency metrics
p50 / p95 / p99
throughput FPS
benchmark JSON
```

## Week 5: Accuracy evaluation

Deliverables:

```text
label parser
IoU calculation
precision / recall
false positive analysis
missed detection examples
```

## Week 6: Report + polish

Deliverables:

```text
HTML or Markdown report
visualizations
README
architecture diagram
performance summary
interview talking points
```

---

# Final README should say something like this

```text
Autonomy-Style Perception Evaluation System

This project implements an end-to-end perception evaluation pipeline for
driving-scene images. It loads a PyTorch object detection model, exports it
to ONNX, runs inference through a C++ inference engine, benchmarks latency
and throughput, and evaluates detection quality using frame-level metrics.

Key features:
- PyTorch model workflow
- ONNX export and validation
- C++ inference runner
- OpenCV preprocessing
- latency / throughput benchmarking
- precision / recall / false-positive evaluation
- Python-generated evaluation reports
```

---

# What this project demonstrates in interviews

You can say:

> “I built a small autonomy-style perception evaluation system. The goal was not to invent a new model, but to understand the production ML infra path: model export, C++ inference, benchmarking, evaluation metrics, and report generation.”

That is a very strong story for:

```text
Waymo simulation/evaluation infra
Tesla Autopilot validation
NVIDIA inference/platform roles
robotics ML infra roles
AV data platform roles
```

This project fits your background especially well because it is not pure ML research. It is **ML systems engineering**.
